File: Week 03/id_113/LeetCode74.py
class Solution:
    # 暴力搜索
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            for i in row:
                if target == i:
                    return True

        return False

    # 降维检索
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        temp = []
        for l in matrix:
            temp.extend(l)

        left = 0
        right = len(temp) - 1

        while left <= right:
            mid = left + (right - left) // 2
            if temp[mid] == target:
                return True
            elif temp[mid] < target:
                left = mid + 1
            else:
                right = mid -1

        return False

    # 不降维检索：
    # 1.从左下角开始，判断target可能在哪一行
    # 2.在该行中进行二分查找
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False

        m, n = len(matrix), len((matrix[0]))

        while matrix[m - 1][0] > target and m > 1:
            m -= 1

        left, right = 0, n - 1
        while left <= right:
            mid = left + (right - left) // 2
            tmp = matrix[m - 1][mid]
            if tmp == target:
                return True
            elif tmp > target:
                right = mid - 1
            else:
                left = mid + 1
        return False

File: Week 03/id_473/LeetCode_74.java
import java.util.ArrayList;
import java.util.List;
/**
 * 74. 搜索二维矩阵
 * @Author CJ
 * @create 2019/11/3
 */
class LeetCode_74 {
    public static void main(String[] args) {
        int[][] matrix = new int[][]{{1,3,5}};
        //int[][] matrix = new int[][]{{},{1}};
        int target = 1;
        System.out.println(searchMatrix(matrix, target));
    }

    //二分查找（二维数组转一维数组）
    public static boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0 || matrix == null) {
            return false;
        }

        Integer[] nums = toOne(matrix);
        if (nums.length == 0 || nums == null){
            return  false;
        }
        int left = 0;
        int right = nums.length - 1;

        while (left + 1 < right) {
            int mid = (left + right)>>>1;
            if (target == nums[mid]) {
                return true;
            }
            if (nums[left] <= target && target <= nums[mid]){
                right = mid;
            }else {
                left = mid;
            }
        }
        if (nums[left] == target || nums[right] == target){
            return true;
        }
        return  false;
    }
    //二维数组转一维数组
    private static Integer[] toOne(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        int i,j = 0;
        for ( i = 0; i < matrix.length; i++) {
            if (matrix[i].length == 0){
                continue;
            }
            for ( j = 0; j < matrix[i].length; j++) {
                list.add(matrix[i][j]);
            }
        }
        return  list.toArray(new Integer[list.size()]);
    }

}